apps/webapp/src/state/swap.ts

```diff
@@ -55,28 +55,38 @@ export const createSwapSlice = (): SliceCreator<SwapSlice> => (set, get) => {
         state.swap.txInProgress = true;
       });
 
-      const { dismiss } = toastFn(loadingTxToast);
+      const { dismiss, update } = toastFn(buildingTxToast());
 
       try {
-        const swapReq = await assembleSwapRequest(get().swap);
-        const swapTx = await planWitnessBuildBroadcast(swapReq);
-        const swapCommitment = getSwapCommitment(swapTx);
+        // build swap
+        const swapPlan = await plan(await assembleSwapRequest(get().swap));
+        const swapTx = await authWitnessBuild(update, { transactionPlan: swapPlan });
 
-        const swapClaimReq = assembleSwapClaimRequest(swapCommitment);
-        const swapClaimTx = await planWitnessBuildBroadcast(swapClaimReq, {
-          /**
-           * Swap claims are self-authenticating, so we'll use `witnessAndBuild`
-           * rather than `authorizeAndBuild` for the swap claim. That way, we
-           * won't trigger a second, unnecessary approval popup.
-           *
-           * @see https://protocol.penumbra.zone/main/zswap/swap.html#claiming-swap-outputs
-           */
-          rpcMethod: 'witnessAndBuild',
+        // begin broadcast
+        const swapBroadcast = broadcast(update, {
+          awaitDetection: true,
+          transaction: swapTx,
         });
-        const swapClaimTxHash = await getTransactionHash(swapClaimTx);
 
-        dismiss();
-        toastFn(successTxToast(swapClaimTxHash));
+        // a swap claim must be broadcast after any successful swap, to claim
+        // the output of the swap.  Swap claims don't need authenticationData,
+        // so use `witnessAndBuild`
+        // @see https://protocol.penumbra.zone/main/zswap/swap.html#claiming-swap-outputs
+
+        // start to build swap claim
+        const swapCommitment = getSwapCommitment(swapTx);
+
+        const swapClaimPlan = await plan(assembleSwapClaimRequest(swapCommitment));
+        const swapClaimTx = await witnessBuild(update, { transactionPlan: swapClaimPlan });
+
+        // after swap broadcast success, emit the claim
+        await swapBroadcast;
```

Comment:
Ohhh smart! allows the build of one to start while the other is still being broadcast 👏🏻